atoma-input-manager/src/lib.rs

```diff
@@ -18,50 +41,110 @@ pub struct AtomaInputManager {
+        info!("Atoma Input Manager started in {:?}", start.elapsed());
+        Ok(Self {
             firebase_input_manager,
             input_manager_rx,
-        }
+            ipfs_request_tx,
+            ipfs_join_handle,
+        })
     }
 
     /// Main loop, responsible for continuously listening to incoming user prompts.
     #[instrument(skip_all)]
     pub async fn run(mut self) -> Result<(), AtomaInputManagerError> {
-        info!("Starting firebase input service..");
+        info!("Starting `AtomaInputManager` service..");
         while let Some((input_source, oneshot)) = self.input_manager_rx.recv().await {
             info!(
                 "Received a new input to be submitted to a data storage {:?}..",
                 input_source
             );
-            let text = match input_source {
+            match input_source {
                 InputSource::Firebase { request_id } => {
-                    self.firebase_input_manager
-                        .handle_get_request(request_id)
-                        .await
+                    let model_input_result = self
+                        .firebase_input_manager
+                        .handle_chat_request(request_id)
+                        .await;
+                    oneshot
+                        .send(model_input_result)
+                        .map_err(|_| AtomaInputManagerError::SendPromptError)?;
+                }
+                InputSource::Ipfs { cid, format } => {
+                    self.ipfs_request_tx
+                        .send((cid, format, oneshot))
+                        .map_err(|_| AtomaInputManagerError::SendPromptError)?;
+                }
+                InputSource::Raw { prompt } => {
+                    oneshot
+                        .send(Ok(ModelInput::Text(prompt)))
+                        .map_err(|_| AtomaInputManagerError::SendPromptError)?;
                 }
-                InputSource::Raw { prompt } => Ok((prompt, vec![])),
-            };
-            oneshot
-                .send(text)
-                .map_err(|_| AtomaInputManagerError::SendPromptError)?;
+            }
         }
+        Ok(())
+    }
+
+    /// Graceful shutdown
+    #[instrument(skip_all)]
+    pub async fn shutdown(self) -> Result<(), AtomaInputManagerError> {
+        info!("Shutting down Atoma Input Manager...");
+
+        trace!("Dropping IPFS request tx...");
+        drop(self.ipfs_request_tx);
+
+        trace!("Aborting IPFS manager join handle...");
+        if let Some(handle) = self.ipfs_join_handle {
+            handle.abort();
+
+            trace!("Waiting for IPFS manager to join...");
+            match handle.await {
+                Ok(_) => Ok(()),
+                Err(e) if e.is_cancelled() => Ok(()),
+                Err(e) => Err(AtomaInputManagerError::JoinError(e)),
+            }?;
+        }
+
+        trace!("Dropping output manager receiver...");
```

Comment:
you are dropping `input_manager_rx` so why output manager receiver?

Reply:
Good catch ! Refactored